talk/concurrency/atomic.tex

```diff
@@ -0,0 +1,95 @@
+\subsection[atomic]{Atomic types}
+
+\begin{frame}[fragile]
+  \frametitlecpp[11]{Atomic types in \cpp}
+  \begin{block}{std::atomic template}
+    \begin{itemize}
+      \item Any trivially copyable type can be made atomic in C++
+      \item Most useful for integral types
+      \item May internally use locks for custom types
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{}
+    \begin{cppcode*}{}
+      std::atomic<int> a{0};
+      std::thread t1([&](){ a++; });
+      std::thread t2([&](){ a++; });
+      a += 2;
+      t1.join(); t2.join();
+      assert( a == 4 ); // Guaranteed to succeed
+    \end{cppcode*}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \begin{alertblock}{Expressions using an atomic type are \textit{not} atomic!}
+    \begin{itemize}
+      \item Atomic load; value+2; atomic store
+    \end{itemize}
+    \begin{cppcode*}{}
+      std::atomic<int> a{0};
+      std::thread t1([&]{ a = a + 2; });
+      std::thread t2([&]{ a = a + 2; });
+    \end{cppcode*}
+  \end{alertblock}
+  \begin{block}{Sequence diagram}
+    \begin{tikzpicture}
+      \begin{umlseqdiag}
+        \umlobject[x=0]{Thread 1}
+        \umlobject[x=3, fill=blue!20]{atomic}
+        \umlobject[x=6]{Thread 2}
+        \begin{umlcall}[op=load]{Thread 1}{atomic}
+        \end{umlcall}
+        \begin{umlcall}[op=+2]{Thread 1}{Thread 1}
+        \end{umlcall}
+        \begin{umlcall}[op=load]{Thread 2}{atomic}
+        \end{umlcall}
+        \begin{umlcall}[op=+2]{Thread 2}{Thread 2}
+        \end{umlcall}
+        \begin{umlcall}[op=store 2]{Thread 1}{atomic}
+        \end{umlcall}
+        \begin{umlcall}[op=store 2]{Thread 2}{atomic}
+        \end{umlcall}
+      \end{umlseqdiag}
+      \draw[-triangle 60](9,0) -- (9,-4) node[right, pos=0.5]{time};
+    \end{tikzpicture}
+  \end{block}
+\end{frame}
+
+\begin{frame}[fragile]
+  \begin{exampleblock}{Use built-in atomic functions}
+    \begin{itemize}
+      \item Atomically \{ load; add; store \}
+    \end{itemize}
+    \begin{cppcode*}{}
+      std::atomic<int> a{0};
+      std::thread t1([&]{ a.fetch_add(2); });
+      std::thread t2([&]{ a.fetch_add(2); });
+    \end{cppcode*}
+  \end{exampleblock}
+  \begin{block}{Sequence diagram}
+    \begin{tikzpicture}
+      \begin{umlseqdiag}
+        \umlobject[x=0]{Thread 1}
+        \umlobject[x=3, fill=blue!20]{atomic}
+        \umlobject[x=6]{Thread 2}
+        \begin{umlcall}[op=fetch\_add]{Thread 1}{atomic}
+        \end{umlcall}
+        \begin{umlcall}[op=fetch\_add, dt=9]{Thread 2}{atomic}
+        \end{umlcall}
+      \end{umlseqdiag}
+      \draw[-triangle 60](9,0) -- (9,-4) node[right, pos=0.5]{time};
```

Comment:
Arrow is too long.
If you shorten it, I would use the new space to mention that `+=` is another name for fetch_add and other operators are also thread safe. And maybe repeat that ` a+= 2` is not at all equivalent to `a = a + 2` in this case.

Reply:
Good point, done!